strawberryfields/circuitspecs/chip0.py

```diff
@@ -51,7 +59,130 @@ class Chip0Specs(CircuitSpecs):
         Rgate({internal_phase_0}) | [2]
         BSgate(pi/4, pi/2) | [2, 3]
 
+        # final local phases
+        Rgate({phi0}) | 0
+        Rgate({phi1}) | 1
+        Rgate({phi2}) | 2
+        Rgate({phi3}) | 3
+
         # Measurement in Fock basis
         MeasureFock() | [0, 1, 2, 3]
         """
     )
+
+    def compile(self, seq, registers):
+        """Try to arrange a quantum circuit into a form suitable for Chip0.
+
+        Args:
+            seq (Sequence[Command]): quantum circuit to modify
+            registers (Sequence[RegRefs]): quantum registers
+        Returns:
+            List[Command]: modified circuit
+        Raises:
+            CircuitError: the circuit does not correspond to Chip0
+        """
+        # pylint: disable=too-many-statements,too-many-branches
+        # First, check if provided sequence matches the circuit template.
+        # This will avoid superfluous compilation if the user is using the
+        # template directly.
+        try:
+            seq = super().compile(seq, registers)
+        except CircuitError:
+            # failed topology check. Continue to more general
+            # compilation below.
+            pass
+        else:
+            return seq
+
+        # first do general GBS compilation to make sure
+        # Fock measurements are correct
+        # ---------------------------------------------
+        seq = GBSSpecs().compile(seq, registers)
+        A, B, C = group_operations(seq, lambda x: isinstance(x, ops.MeasureFock))
+
+        if len(B[0].reg) != self.modes:
+            raise CircuitError('All modes must be measured.')
```

Comment:
Potentially, but does this make sense with hardware? It would mean throwing out results before returning it to the user.

With most qubit frameworks, all modes are assumed to terminate with measurement, even if the user doesn't specifically place a measurement operation (I like this approach). Some (Q#) even force all modes to be measured before allowing the code to compile (potentially too strict).

Reply:
In the near term, the hardware detector will perform a a measurement anyways. Not collecting the measurement result from a particular detection channel would of course reduce the network traffic and result memory footprint, so if there is enough time to implement this feature, it is not pointless (even though I do not see the use-case). In the more distant future, I can come up with two ideas for this:
- If detection channels should be a limiting factor some day (opposed to available chips), and an optical multiplexer would distribute the chip output modes among available detection channels, one could actually benefit from this as the output from more chips could be measured simultaneously. However, adding a multiplexer introduces loss and I therefore doubt this will ever be a useful scenario, as opposed to having a fixed (spliced) routing from chip output modes to detector channels.
- Maybe this makes sense if a circuit is expected to be "plugged" into another circuit one day?

Reply:
Let's make this an issue, as at the moment SF does not have the capability to specify a topology with optional gates